File: projects/enwiki9/tools/midas_open_boundary_observer_v1.py
#!/usr/bin/env python3
"""Bounded observer successor build and streamed first-divergence diagnostics.

The codec law and v1 sources remain unchanged. SHA-256 witnesses cover serialized
bytes only; fixture snapshots permit direct equality and independent coverage
checks. Neither this wrapper nor a matching digest grants a corpus certificate.
"""
from __future__ import annotations

import argparse
import contextlib
import hashlib
import json
import os
from pathlib import Path
import stat
import struct
import subprocess
import sys
import tempfile
import time

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))
from tools import midas_open_codec_v1 as parent

FILES = ("data", "state.bin", "summary.json", "probabilities.bin", "boundaries.jsonl", "snapshots.bin")
FILE_LIMIT = 32 * 1024**2
PART_NAMES = (
    "complete_state", "complete_predictor", "parent_identity_projection", "normalized_coder",
    "reference_model_projection", "scheduler", "byte_prefix", "decoded_prefix", "sequence_origin",
    "scheduler_update_counters", "discarded_shadow", "model_header_prefix", "model_probability_cache",
    "recurrent_memory", "parameters", "optimizer_moments_and_compensation", "incremental_cache",
)
SOURCES = (ROOT / "tools/midas_open_boundary_observer_v1.cpp", *parent.SOURCES[1:],
           parent.CORE / "profile_population.cpp")


def build(cache_dir: Path):
    return parent.build_cpp_cached(sources=SOURCES, flags=parent.FLAGS,
                                   cache_dir=cache_dir, timeout_seconds=120)


@contextlib.contextmanager
def bounded_input(path: Path):
    fd = os.open(path, os.O_RDONLY | os.O_NOFOLLOW | os.O_NONBLOCK | os.O_CLOEXEC)
    with os.fdopen(fd, "rb") as source:
        before = os.fstat(source.fileno())
        if not stat.S_ISREG(before.st_mode) or not 0 <= before.st_size <= FILE_LIMIT:
            raise ValueError(f"observer input is not a bounded regular file: {path}")
        yield source
        after, current = os.fstat(source.fileno()), path.lstat()
        identity = lambda row: (row.st_dev, row.st_ino, row.st_size, row.st_mtime_ns, row.st_ctime_ns)
        if identity(before) != identity(after) or identity(after) != identity(current):
            raise ValueError(f"observer input changed while read: {path}")


def probability_header(source) -> dict:
    header = source.read(21)
    if len(header) != 21 or header[:8] != b"MOPROB01":
        raise ValueError("probability trace header missing or malformed")
    raw, model, arm = struct.unpack("<QIB", header[8:])
    if raw > 250000 or model != 0x4F504601 or arm > 2:
        raise ValueError("probability trace coordinates/model/arm invalid")
    actual = os.fstat(source.fileno()).st_size
    if actual != 21 + 16 * raw:
        raise ValueError(f"probability trace incomplete: observed {actual} bytes, expected {21 + 16 * raw}")
    return {"raw_bytes": raw, "model_tag": model, "archive_arm": arm, "records": raw * 8}


def boundary_rows(path: Path, raw_bytes: int):
    expected = [("initial", 0), *(("boundary", n * 8) for n in range(32, raw_bytes + 1, 32)),
                ("final", raw_bytes * 8)]
    with bounded_input(path) as source:
        for kind, position in expected:
            line = source.readline(8193)
            if not line.endswith(b"\n") or len(line) > 8192:
                raise ValueError(f"boundary trace missing/truncated at {kind}:{position}")
            row = json.loads(line)
            if row.get("kind") != kind or row.get("bit_position") != position:
                raise ValueError(f"boundary coordinates differ at {kind}:{position}")
            if [part.get("name") for part in row.get("parts", [])] != list(PART_NAMES):
                raise ValueError("boundary state component population differs")
            for part in row["parts"]:
                if (type(part.get("offset")) is not int or type(part.get("bytes")) is not int or
                        part["offset"] < 0 or part["bytes"] < 0 or
                        part["offset"] + part["bytes"] > row["parts"][0]["bytes"] or
                        not isinstance(part.get("sha256"), str) or len(part["sha256"]) != 64 or
                        any(c not in "0123456789abcdef" for c in part["sha256"])):
                    raise ValueError("boundary component range/digest invalid")
            if any(type(row.get(name)) is not int or row[name] < 0 for name in (
                    "model_updates", "parent_updates", "midpoint_updates", "shadow_updates")):
                raise ValueError("boundary update counters invalid")
            yield row
        if source.read(1):
            raise ValueError("boundary trace has extra records")


def snapshot_records(path: Path):
    with bounded_input(path) as source:
        header = source.read(9)
        if header not in (b"MOSNAP01\0", b"MOSNAP01\1"):
            raise ValueError("snapshot header differs")
        if header[-1] == 0:
            if source.read(1):
                raise ValueError("disabled snapshots contain records")
            return
        for _ in range(6):  # initial + four 32-byte boundaries + final at <=129 bytes
            header = source.read(17)
            if not header:
                return
            if len(header) != 17:
                raise ValueError("truncated snapshot record header")
            kind, bit, size = struct.unpack("<BQQ", header)
            if kind > 2 or bit > 129 * 8 or size > 8 * 1024**2:
                raise ValueError("snapshot record exceeds fixture bounds")
            state = source.read(size)
            if len(state) != size:
                raise ValueError("snapshot record truncated")
            yield {"kind": ("initial", "boundary", "final")[kind], "bit_position": bit, "state": state}
        if source.read(1):
            raise ValueError("snapshot population exceeds fixture bound")


def validate_snapshot(row: dict, snapshot: dict) -> None:
    if (row["kind"], row["bit_position"]) != (snapshot["kind"], snapshot["bit_position"]):
        raise ValueError("snapshot boundary differs")
    state = snapshot["state"]
    if len(state) != row["parts"][0]["bytes"] or state[:5] != b"GMST\1":
        raise ValueError("snapshot envelope differs")
    for part in row["parts"]:
        block = state[part["offset"]:part["offset"] + part["bytes"]]
        if len(block) != part["bytes"] or hashlib.sha256(block).hexdigest() != part["sha256"]:
            raise ValueError("snapshot digest differs: " + part["name"])


def validate_bundle(directory: Path) -> dict:
    with bounded_input(directory / "summary.json") as source:
        summary = json.load(source)
    if summary.get("schema") != "midas_open_boundary_observer_v1":
        raise ValueError("observer summary schema differs")
    with bounded_input(directory / "probabilities.bin") as source:
        header = probability_header(source)
        while chunk := source.read(65536):
            if any(value == 0 for (value,) in struct.iter_unpack("<H", chunk)):
                raise ValueError("probability trace contains invalid Q16 zero")
    if summary["raw_bytes"] != header["raw_bytes"] or summary["probability_records"] != header["records"]:
        raise ValueError("summary probability counts differ")
    arm = summary.get("arm")
    if arm not in "PKFS" or len(arm) != 1 or header["archive_arm"] != (1 if arm == "F" else 2 if arm == "S" else 0):
        raise ValueError("summary arm differs from probability law")
    snapshots = snapshot_records(directory / "snapshots.bin")
    count = 0
    for row in boundary_rows(directory / "boundaries.jsonl", header["raw_bytes"]):
        n = row["bit_position"] // 8
        midpoint = n // 64 + int(n % 64 >= 32)
        expected = (n // 64 + (midpoint if arm in "FS" else 0), n // 64,
                    midpoint if arm in "FS" else 0, midpoint if arm == "K" else 0)
        if tuple(row[name] for name in ("model_updates", "parent_updates", "midpoint_updates", "shadow_updates")) != expected:
            raise ValueError("boundary counters differ from causal schedule")
        if summary["exact_snapshots"]:
            snapshot = next(snapshots, None)
            if snapshot is None:
                raise ValueError("mandatory exact snapshot missing")
            validate_snapshot(row, snapshot)
        count += 1
    if next(snapshots, None) is not None:
        raise ValueError("unexpected extra snapshot")
    if count != summary["boundary_records"]:
        raise ValueError("summary boundary count differs")
    records = {name: parent.file_record(directory / name) for name in FILES}
    for name, key in (("probabilities.bin", "probability_bytes"), ("boundaries.jsonl", "boundary_bytes"),
                      ("snapshots.bin", "snapshot_bytes"), ("state.bin", "state_bytes")):
        if records[name]["bytes"] != summary[key]:
            raise ValueError("summary artifact bytes differ: " + name)
    with bounded_input(directory / "state.bin") as source:
        final_state = source.read()
    if hashlib.sha256(final_state).hexdigest() != row["parts"][0]["sha256"]:
        raise ValueError("final boundary differs from final state")
    return {"summary": summary, "artifacts": records, "header": header}


def execute(built, *, operation: str, arm: str, max_raw_bytes: int, source: Path,
            output: Path, wall_seconds: int, snapshots: bool = False) -> dict:
    if operation not in ("encode", "decode") or arm not in tuple("PKFS"):
        raise ValueError("invalid codec operation/arm")
    if not 1 <= max_raw_bytes <= 250000 or not 1 <= wall_seconds <= 120:
        raise ValueError("explicit raw/wall bounds exceed observer limits")
    before = parent.verified_binary(built)
    command = [str(built.binary), operation, arm, str(max_raw_bytes), str(source), str(output),
               "snapshots" if snapshots else "digest"]
    start = time.monotonic()
    result = subprocess.run(command, capture_output=True, text=True, timeout=wall_seconds,
                            env={"PATH": os.defpath, "LC_ALL": "C", "OMP_NUM_THREADS": "1", "OPENBLAS_NUM_THREADS": "1"})
    if parent.file_record(built.binary) != before:
        raise ValueError("observer executable changed during operation")
    if result.returncode:
        raise ValueError(f"observer process failed rc={result.returncode}: {result.stderr.strip()}")
    bundle = validate_bundle(output)
    if bundle["summary"] != json.loads(result.stdout):
        raise ValueError("observer stdout differs from published summary")
    return {"schema": "midas_open_boundary_observer_execution_v1", "bundle": bundle,
            "binary": before, "external_wall_seconds": time.monotonic() - start,
            "timing_scope": "shared-host whole codec including observation and publication; not kernel-only",
            "resource_qualified": False, "complete_package_bytes": None, "objective_credit_bytes": 0}


def _context(source, record: int, total: int) -> dict:
    start, end = max(0, record - 4), min(total, record + 5)
    source.seek(21 + 2 * start)
    return {"first_bit": start, "probabilities": [x[0] for x in struct.iter_unpack("<H", source.read(2 * (end - start)))]}


def _comparison(reference: Path, target: Path, projection: str) -> dict:
    if projection not in ("complete", "parent"):
        raise ValueError("comparison projection must be complete or parent")
    with bounded_input(reference / "probabilities.bin") as a, bounded_input(target / "probabilities.bin") as b:
        ah, bh = probability_header(a), probability_header(b)
        if ah != bh:
            return {"equal": False, "kind": "probability-identity", "reference": ah, "target": bh}
        offset = 0
        while left := a.read(65536):
            right = b.read(len(left))
            if left != right:
                first = next(i for i, pair in enumerate(zip(left, right)) if pair[0] != pair[1])
                bit = (offset + first) // 2
                return {"equal": False, "kind": "probability", "bit_position": bit,
                        "reference_context": _context(a, bit, ah["records"]),
                        "target_context": _context(b, bit, ah["records"])}
            offset += len(left)
    a_rows = boundary_rows(reference / "boundaries.jsonl", ah["raw_bytes"])
    b_rows = boundary_rows(target / "boundaries.jsonl", ah["raw_bytes"])
    previous = None
    while True:
        a, b = next(a_rows, None), next(b_rows, None)
        if a is None or b is None:
            if a is not b:
                raise ValueError("boundary population differs")
            break
        names = PART_NAMES if projection == "complete" else ("parent_identity_projection", "normalized_coder")
        ap, bp = ({p["name"]: (p["bytes"], p["sha256"]) for p in row["parts"]} for row in (a, b))
        differing = [name for name in names if ap[name] != bp[name]]
        if differing:
            after = {"reference": next(a_rows, None), "target": next(b_rows, None)}
            # Context contains actual adjacent boundary records; hashes are not
            # expanded into invented detailed state. Exact snapshots are separate.
            return {"equal": False, "kind": "boundary", "bit_position": a["bit_position"],
                    "boundary_kind": a["kind"], "components": differing, "previous": previous,
                    "current": {"reference": a, "target": b}, "next": after}
        previous = {"reference": a, "target": b}
    return {"equal": True, "probability_records": ah["records"],
            "boundary_records": ah["raw_bytes"] // 32 + 2,
            "claim_scope": "all pre-truth Q16 records and selected serialized-state SHA-256 witnesses; no unobserved state claim"}


def compare(reference: Path, target: Path, *, projection: str = "complete", diagnostic: Path | None = None) -> dict:
    try:
        result = _comparison(reference, target, projection)
        if result["equal"]:
            # Equality of two malformed traces cannot certify synchronization.
            # Validate both closed bundles before publishing a passing result;
            # preserve first-divergence context when their traces already differ.
            validate_bundle(reference)
            validate_bundle(target)
    except (OSError, ValueError, KeyError, TypeError) as error:
        result = {"equal": False, "kind": "trace-completeness", "error": str(error)}
    result.update({"schema": "midas_open_boundary_comparison_v1", "reference_path": str(reference),
                   "target_path": str(target), "projection": projection, "objective_credit_bytes": 0})
    if diagnostic is not None:
        data = (json.dumps(result, sort_keys=True) + "\n").encode()
        if len(data) > 64 * 1024:
            raise ValueError("first-divergence diagnostic exceeds byte ceiling")
        # Exclusive atomic link publication also rejects an existing symlink.
        fd, name = tempfile.mkstemp(prefix=".midas-comparison-", dir=diagnostic.parent)
        try:
            with os.fdopen(fd, "wb") as stream:
                stream.write(data); stream.flush(); os.fsync(stream.fileno())
            os.link(name, diagnostic)
        finally:
            os.unlink(name)
    return result


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    sub = parser.add_subparsers(dest="action", required=True)
    for action in ("build", "encode", "decode"):
        command = sub.add_parser(action)
        command.add_argument("--cache-dir", type=Path, required=True)
        if action != "build":
            command.add_argument("--arm", choices=tuple("PKFS"), required=True)
            command.add_argument("--max-raw-bytes", type=int, required=True)
            command.add_argument("--wall-seconds", type=int, required=True)
            command.add_argument("--input", type=Path, required=True)
            command.add_argument("--output-dir", type=Path, required=True)
            command.add_argument("--snapshots", action="store_true")
    command = sub.add_parser("compare")
    command.add_argument("--reference", type=Path, required=True)
    command.add_argument("--target", type=Path, required=True)
    command.add_argument("--projection", choices=("complete", "parent"), default="complete")
    command.add_argument("--diagnostic", type=Path, required=True)
    args = parser.parse_args()
    try:
        if args.action == "compare":
            result = compare(args.reference, args.target, projection=args.projection, diagnostic=args.diagnostic)
        else:
            built = build(args.cache_dir)
            result = ({"binary": str(built.binary), "manifest": built.manifest, "cache_hit": built.cache_hit}
                      if args.action == "build" else execute(
                          built, operation=args.action, arm=args.arm, max_raw_bytes=args.max_raw_bytes,
                          source=args.input, output=args.output_dir, wall_seconds=args.wall_seconds,
                          snapshots=args.snapshots))
        print(json.dumps(result, sort_keys=True))
        return 0 if result.get("equal", True) else 1
    except (OSError, ValueError, parent.BuildCacheError, subprocess.SubprocessError) as error:
        print("midas_open_boundary_observer_v1: " + str(error), file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: projects/enwiki9/tools/midas_open_boundary_observer_v1.cpp
// Observer successor only: the included, measured codec and model law are immutable.
// Archives retain v1 framing. Observation files are diagnostics, not codec inputs.
#define main midas_open_unobserved_main_v1
#include "midas_open_codec_v1.cpp"
#undef main
#include "../programs/nncp_open_integrated_midpoint_segment_replay_65536_q0_v2/profile_population.hpp"

#include <array>
#include <span>

namespace {
constexpr std::size_t observer_file_limit = 32 * 1024 * 1024;
constexpr std::size_t snapshot_raw_limit = 129;
constexpr const char* observer_error = "midas_open_boundary_observer_v1: ";

// Anonymous, already-unlinked files stream during prediction. A process stop
// cannot leave a named probability/state trace mistaken for a complete result.
class ObserverStream {
 public:
  explicit ObserverStream(const fs::path& parent) {
    auto pattern = (parent / ".midas-observer-stream-XXXXXX").string();
    fd_ = ::mkstemp(pattern.data());
    require(fd_ >= 0, "cannot create private observer stream");
    if (::unlink(pattern.c_str()) != 0) {
      ::close(fd_); fd_ = -1;
      throw std::runtime_error("cannot unlink private observer stream");
    }
    require(::fcntl(fd_, F_SETFD, FD_CLOEXEC) == 0, "cannot protect observer descriptor");
  }
  ObserverStream(const ObserverStream&) = delete;
  ~ObserverStream() { if (fd_ >= 0) ::close(fd_); }
  void write(std::span<const std::uint8_t> bytes) {
    require(bytes.size() <= observer_file_limit - bytes_, "observer file exceeds byte ceiling");
    std::size_t offset = 0;
    while (offset < bytes.size()) {
      const auto n = ::write(fd_, bytes.data() + offset, bytes.size() - offset);
      if (n < 0 && errno == EINTR) continue;
      require(n > 0, "observer stream write failed"); offset += n;
    }
    bytes_ += bytes.size();
  }
  void text(const std::string& value) {
    write(std::span(reinterpret_cast<const std::uint8_t*>(value.data()), value.size()));
  }
  void publish(const fs::path& path) {
    require(::fsync(fd_) == 0 && ::lseek(fd_, 0, SEEK_SET) == 0, "observer stream sync failed");
    File target(::open(path.c_str(), O_WRONLY | O_CREAT | O_EXCL | O_NOFOLLOW | O_CLOEXEC, 0600));
    std::array<std::uint8_t, 65536> buffer{};
    std::size_t copied = 0;
    while (copied < bytes_) {
      const auto n = ::read(fd_, buffer.data(), std::min(buffer.size(), bytes_ - copied));
      if (n < 0 && errno == EINTR) continue;
      require(n > 0, "observer stream copy read failed");
      std::size_t offset = 0;
      while (offset < std::size_t(n)) {
        const auto m = ::write(target.descriptor, buffer.data() + offset, n - offset);
        if (m < 0 && errno == EINTR) continue;
        require(m > 0, "observer stream copy write failed"); offset += m;
      }
      copied += n;
    }
    require(::fsync(target.descriptor) == 0, "observer output sync failed");
    const auto fd = target.descriptor; target.descriptor = -1;
    require(::close(fd) == 0, "observer output close failed");
  }
  std::size_t size() const { return bytes_; }
 private:
  int fd_ = -1;
  std::size_t bytes_ = 0;
};

struct StatePart { std::string name; std::size_t offset, bytes; };
// Parse only the versioned serialization produced by the unchanged classes.
// Ranges point into the complete snapshot; no state field is manufactured here.
class StateCursor {
 public:
  StateCursor(const Bytes& bytes, std::size_t begin, std::size_t size)
      : data_(bytes), position(begin), end_(begin + size) {
    require(begin <= bytes.size() && size <= bytes.size() - begin, "observer state range exceeds snapshot");
  }
  void skip(std::size_t count) {
    require(count <= end_ - position, "observer state range truncated"); position += count;
  }
  std::uint64_t integer(unsigned width) {
    require(width <= 8 && width <= end_ - position, "observer integer truncated");
    std::uint64_t value = 0;
    for (unsigned i = 0; i < width; ++i) value |= std::uint64_t(data_[position++]) << (8 * i);
    return value;
  }
  StatePart blob(const std::string& name) {
    const auto count = integer(8), begin = position; skip(count); return {name, begin, count};
  }
  void vector(unsigned width) {
    const auto count = integer(8);
    require(count <= (end_ - position) / width, "observer vector truncated"); skip(count * width);
  }
  void magic(const char* expected) {
    require(end_ - position >= 5 && std::memcmp(data_.data() + position, expected, 5) == 0,
            "observer serialization version differs"); skip(5);
  }
  void finish() const { require(position == end_, "observer serialization has unparsed bytes"); }
  const Bytes& data_;
  std::size_t position;
 private:
  std::size_t end_;
};

std::vector<StatePart> state_parts(const Bytes& state, const Predictor& model) {
  std::vector<StatePart> parts{{"complete_state", 0, state.size()}};
  StateCursor envelope(state, 0, state.size()); envelope.magic("GMST\1");
  for (const auto name : {"complete_predictor", "parent_identity_projection", "normalized_coder", "reference_model_projection"})
    parts.push_back(envelope.blob(name));
  envelope.finish();
  StateCursor predictor(state, parts[1].offset, parts[1].bytes); predictor.magic("MBIT\1");
  const auto schedule = predictor.blob("scheduler"), prefix = predictor.blob("byte_prefix");
  predictor.finish(); parts.push_back(schedule); parts.push_back(prefix);
  StateCursor scheduler(state, schedule.offset, schedule.bytes); scheduler.magic("MSCH\1");
  scheduler.skip(1 + 8 + 8 + 1 + 256 * 4);
  parts.push_back(scheduler.blob("decoded_prefix"));
  parts.push_back(scheduler.blob("sequence_origin"));
  const auto backend = scheduler.blob("backend_model");
  // The scheduler digest above also includes arm, update counters and K shadow.
  parts.push_back({"scheduler_update_counters", scheduler.position, 16}); scheduler.skip(16);
  parts.push_back(scheduler.blob("discarded_shadow")); scheduler.finish();
  StateCursor native(state, backend.offset, backend.bytes); native.magic("OPFI\1");
  native.skip(4 + 8 + 1 + 3); const auto prefix_bytes = native.integer(1); native.skip(prefix_bytes);
  parts.push_back({"model_header_prefix", backend.offset, native.position - backend.offset});
  const auto cache_begin = native.position; native.skip(256 * 4);
  parts.push_back({"model_probability_cache", cache_begin, native.position - cache_begin});
  const auto memory_begin = native.position;
  for (std::size_t i = 0; i < Model::geometry().layers; ++i) native.vector(2);
  parts.push_back({"recurrent_memory", memory_begin, native.position - memory_begin});
  const auto parameters_begin = native.position;
  const auto parameters = model.schedule().model().parameter_states();
  for (const auto& parameter : parameters) {
    const auto begin = native.position; native.skip(parameter.second.size());
    require(std::equal(parameter.second.begin(), parameter.second.end(), state.begin() + begin),
            "observer parameter range differs from named model snapshot");
  }
  parts.push_back({"parameters", parameters_begin, native.position - parameters_begin});
  const auto optimizer_begin = native.position; native.skip(8);
  for (std::size_t i = 0; i < parameters.size(); ++i) {
    native.vector(2); native.vector(2); native.vector(4);
  }
  parts.push_back({"optimizer_moments_and_compensation", optimizer_begin, native.position - optimizer_begin});
  parts.push_back(native.blob("incremental_cache")); native.finish();
  return parts;
}

class BoundaryObserver {
 public:
  BoundaryObserver(const fs::path& parent, std::size_t raw_bytes, unsigned arm, bool snapshots)
      : probabilities(parent), boundaries(parent), states(parent), snapshots_(snapshots) {
    require(!snapshots || raw_bytes <= snapshot_raw_limit, "exact boundary snapshots require at most 129 raw bytes");
    Bytes header{'M','O','P','R','O','B','0','1'}; put(header, raw_bytes, 8);
    put(header, Model::model_tag, 4); put(header, arm, 1); probabilities.write(header);
    Bytes snapshot_header{'M','O','S','N','A','P','0','1'}; put(snapshot_header, snapshots, 1);
    states.write(snapshot_header);
  }
  void probability(std::uint16_t value) {
    // Called after predict(), before encoder truth access / decoder.decode().
    require(value > 0, "observer encountered zero Q16 probability");
    buffer_.push_back(value & 255); buffer_.push_back(value >> 8); ++bits_;
    if (buffer_.size() == 65536) flush();
  }
  void boundary(const char* kind, unsigned tag, const Predictor& model, const Bytes& coder) {
    require(model.bit_position() == bits_, "observer probability clock differs from model");
    const auto state = complete_state(model, coder); const auto parts = state_parts(state, model);
    std::ostringstream row;
    row << "{\"kind\":\"" << kind << "\",\"bit_position\":" << bits_
        << ",\"model_updates\":" << model.schedule().model().updates()
        << ",\"parent_updates\":" << model.schedule().parent_updates()
        << ",\"midpoint_updates\":" << model.schedule().midpoint_updates()
        << ",\"shadow_updates\":" << model.schedule().shadow_updates() << ",\"parts\":[";
    bool first = true;
    for (const auto& part : parts) {
      if (!first) row << ',';
      first = false;
      row << "{\"name\":\"" << part.name << "\",\"offset\":" << part.offset << ",\"bytes\":" << part.bytes
          << ",\"sha256\":\"" << gamma_enwiki9::nncp::Sha256Hex(
               std::span(state.data() + part.offset, part.bytes)) << "\"}";
    }
    row << "]}\n"; boundaries.text(row.str());
    if (snapshots_) {
      Bytes header; put(header, tag, 1); put(header, bits_, 8); put(header, state.size(), 8);
      states.write(header); states.write(state);
    }
    ++records_;
  }
  void flush() { probabilities.write(buffer_); buffer_.clear(); }
  std::uint64_t records() const { return records_; }
  ObserverStream probabilities, boundaries, states;
 private:
  bool snapshots_;
  Bytes buffer_;
  std::uint64_t bits_ = 0, records_ = 0;
};

class ObservedBundle {
 public:
  explicit ObservedBundle(const fs::path& destination) : destination_(destination) {
    check_destination(destination);
    auto pattern = (destination.parent_path() / ".midas-observed-XXXXXX").string();
    const auto name = ::mkdtemp(pattern.data()); require(name != nullptr, "cannot create observed output staging");
    staging_ = name;
  }
  ~ObservedBundle() {
    if (staging_.empty()) return;
    for (const auto name : {"data", "state.bin", "summary.json", "probabilities.bin", "boundaries.jsonl", "snapshots.bin"})
      ::unlink((staging_ / name).c_str());
    ::rmdir(staging_.c_str());
  }
  void publish(const Bytes& output, const Bytes& state, const std::string& summary, BoundaryObserver& observer) {
    // All streams and ordinary outputs use the same checked sync/close path.
    for (const auto& item : std::array<std::pair<const char*, const Bytes*>, 2>{{{"data", &output}, {"state.bin", &state}}}) {
      ObserverStream stream(destination_.parent_path()); stream.write(*item.second); stream.publish(staging_ / item.first);
    }
    ObserverStream summary_stream(destination_.parent_path()); summary_stream.text(summary);
    summary_stream.publish(staging_ / "summary.json");
    observer.probabilities.publish(staging_ / "probabilities.bin");
    observer.boundaries.publish(staging_ / "boundaries.jsonl");
    observer.states.publish(staging_ / "snapshots.bin");
    File directory(::open(staging_.c_str(), O_RDONLY | O_DIRECTORY | O_CLOEXEC));
    require(::fsync(directory.descriptor) == 0, "observed directory sync failed");
    require(::syscall(SYS_renameat2, AT_FDCWD, staging_.c_str(), AT_FDCWD, destination_.c_str(),
                      1U /* RENAME_NOREPLACE */) == 0, "observed output publication refused");
    staging_.clear();
  }
 private:
  fs::path destination_, staging_;
};
}  // namespace

int main(int argc, char** argv) {
  try {
    require(argc == 7, "usage: midas-boundary-observer encode/decode P/K/F/S MAX_RAW_BYTES INPUT NEW_OUTPUT_DIRECTORY digest/snapshots");
    const std::string operation = argv[1], observation = argv[6];
    require(operation == "encode" || operation == "decode", "operation must be encode or decode");
    require(observation == "digest" || observation == "snapshots", "observation must be digest or snapshots");
    const auto arm = parse_arm(argv[2]); const auto limit = raw_limit(argv[3]);
    const auto archive_limit = 32 * limit + 64;
    set_bound(RLIMIT_AS, memory_limit); set_bound(RLIMIT_CPU, 120); set_bound(RLIMIT_FSIZE, observer_file_limit);
    const auto destination = fs::absolute(argv[5]); check_destination(destination);
    const auto start = std::chrono::steady_clock::now(); const auto cpu = std::clock();
    const auto input = read_input(argv[4], operation == "encode" ? limit : archive_limit);
    const auto archive_arm = arm == Arm::F ? 1 : arm == Arm::S ? 2 : 0;
    // Unframe before allocating observers so rejected headers cannot publish diagnostics.
    const auto raw_bytes = operation == "encode" ? input.size() : unframe(input, Model::model_tag, limit, archive_limit).raw_bytes;
    BoundaryObserver observer(destination.parent_path(), raw_bytes, archive_arm, observation == "snapshots");
    Predictor model(Model{}, arm); Bytes raw, output, state;
    if (operation == "encode") {
      raw = input; Encoder coder;
      observer.boundary("initial", 0, model, coder.comparison_state());
      for (auto byte : raw) for (int shift = 7; shift >= 0; --shift) {
        const auto probability = model.predict(); observer.probability(probability);
        const auto truth = std::uint16_t((byte >> shift) & 1U);
        coder.encode(truth, probability); model.observe(truth);
        if (model.bit_position() % 256 == 0) observer.boundary("boundary", 1, model, coder.comparison_state());
      }
      output = frame(raw, coder.finish(), Model::model_tag, archive_arm);
      require(output.size() <= archive_limit, "encoded archive exceeds operational ceiling");
      observer.boundary("final", 2, model, coder.comparison_state());
      state = complete_state(model, coder.comparison_state());
    } else {
      const auto framed = unframe(input, Model::model_tag, limit, archive_limit);
      require(framed.arm == archive_arm, "archive arm differs from decoder law");
      Decoder coder(framed.payload); raw.reserve(framed.raw_bytes);
      observer.boundary("initial", 0, model, coder.comparison_state());
      for (std::size_t i = 0; i < framed.raw_bytes; ++i) {
        unsigned byte = 0;
        for (unsigned bit = 0; bit != 8; ++bit) {
          const auto probability = model.predict(); observer.probability(probability);
          const auto truth = coder.decode(probability); model.observe(truth); byte = (byte << 1) | truth;
          if (model.bit_position() % 256 == 0) observer.boundary("boundary", 1, model, coder.comparison_state());
        }
        raw.push_back(static_cast<std::uint8_t>(byte));
      }
      coder.finish(); verify_inverse(framed, raw);
      observer.boundary("final", 2, model, coder.comparison_state());
      output = raw; state = complete_state(model, coder.comparison_state());
    }
    observer.flush();
    struct rusage usage{}; require(::getrusage(RUSAGE_SELF, &usage) == 0, "resource observation failed");
    std::ostringstream summary; summary.precision(12);
    summary << "{\"schema\":\"midas_open_boundary_observer_v1\",\"operation\":\"" << operation
            << "\",\"arm\":\"" << argv[2] << "\",\"frontend\":\"raw_identity_v1\",\"raw_bytes\":" << raw.size()
            << ",\"archive_bytes\":" << (operation == "encode" ? output.size() : input.size())
            << ",\"state_bytes\":" << state.size() << ",\"max_raw_bytes\":" << limit
            << ",\"probability_records\":" << raw.size() * 8 << ",\"boundary_records\":" << observer.records()
            << ",\"probability_bytes\":" << observer.probabilities.size() << ",\"boundary_bytes\":" << observer.boundaries.size()
            << ",\"snapshot_bytes\":" << observer.states.size() << ",\"exact_snapshots\":" << (observation == "snapshots" ? "true" : "false")
            << ",\"model_updates\":" << model.schedule().model().updates() << ",\"process_peak_rss_kib\":" << usage.ru_maxrss
            << ",\"observed_codec_cpu_seconds\":" << double(std::clock() - cpu) / CLOCKS_PER_SEC
            << ",\"observed_codec_wall_seconds\":" << std::chrono::duration<double>(std::chrono::steady_clock::now() - start).count()
            << ",\"timing_scope\":\"whole codec including observer; excludes final publication; shared-host diagnostic\""
            << ",\"resource_qualified\":false,\"complete_package_bytes\":null,\"objective_credit_bytes\":0}\n";
    ObservedBundle bundle(destination); bundle.publish(output, state, summary.str(), observer);
    std::cout << summary.str(); require(bool(std::cout), "observer stdout publication failed");
    return 0;
  } catch (const std::exception& error) {
    std::cerr << observer_error << error.what() << '\n'; return 1;
  }
}
